scripts/common/upgrade.sh

```diff
@@ -1,186 +1,395 @@
 }
 
-function upgrade_kubernetes_step() {
-    if [ "$KUBERNETES_STEP_LOCAL_PRIMARY" == "1" ]; then
-        upgrade_kubernetes_local_master_minor "$STEP_VERSION"
-    fi
-    if [ "$KUBERNETES_STEP_REMOTE_PRIMARIES" == "1" ]; then
-        upgrade_kubernetes_remote_masters_minor "$STEP_VERSION"
-    fi
-    if [ "$KUBERNETES_STEP_SECONDARIES" == "1" ]; then
-        upgrade_kubernetes_workers_minor "$STEP_VERSION"
-    fi
+# kubernetes_upgrade upgrades will fetch the add-on and load the images for the upgrade and finally
+# run the upgrade script.
+function kubernetes_upgrade() {
+    local from_version="$1"
+    local to_version="$2"
+
+    disable_rook_ceph_operator
+
+    # when invoked in a subprocess the failure of this function will not cause the script to exit
+    # sanity check that the version is valid
+    common_upgrade_step_versions "STEP_VERSIONS[@]" "$from_version" "$to_version" 1>/dev/null
+
+    logStep "Upgrading Kubernetes from $from_version.x to $to_version.x"
+    common_upgrade_print_list_of_minor_upgrades "$from_version" "$to_version"
+    echo "This may take some time."
+    kubernetes_upgrade_addon_fetch "$from_version" "$to_version"
+
+    kubernetes_upgrade_prompt_missing_images "$from_version" "$to_version"
+
+    kubernetes_upgrade_do_kubernetes_upgrade "$from_version" "$to_version"
+
+    enable_rook_ceph_operator
+
+    logSuccess "Successfully upgraded Kubernetes from $from_version.x to $to_version.x"
 }
 
-function upgrade_kubernetes_minor() {
-    if [ "$KUBERNETES_UPGRADE_LOCAL_PRIMARY_MINOR" == "1" ]; then
-        upgrade_kubernetes_local_master_minor "$KUBERNETES_VERSION"
-    fi
-    if [ "$KUBERNETES_UPGRADE_REMOTE_PRIMARIES_MINOR" == "1" ]; then
-        upgrade_kubernetes_remote_masters_minor "$KUBERNETES_VERSION"
-    fi
-    if [ "$KUBERNETES_UPGRADE_SECONDARIES_MINOR" == "1" ]; then
-        upgrade_kubernetes_workers_minor "$KUBERNETES_VERSION"
+# kubernetes_upgrade_do_kubernetes_upgrade will step through each minor version upgrade from $from_version to
+# $to_version
+function kubernetes_upgrade_do_kubernetes_upgrade() {
+    local from_version="$1"
+    local to_version="$2"
+
+    local step=
+    while read -r step; do
+        if [ -z "$step" ] || [ "$step" = "0.0.0" ]; then
+            continue
+        fi
+        if [ ! -d "$DIR/packages/kubernetes/$step" ] ; then
+            bail "Kubernetes version $step not found"
+        fi
+        logStep "Upgrading cluster to Kubernetes version $step"
+        
+        if [ "$KUBERNETES_UPGRADE_LOCAL_PRIMARY" == "1" ]; then
+            upgrade_kubernetes_local_master "$step"
+        fi
+        if [ "$KUBERNETES_UPGRADE_REMOTE_PRIMARIES" == "1" ]; then
+            upgrade_kubernetes_remote_masters "$step"
+        fi
+        if [ "$KUBERNETES_UPGRADE_SECONDARIES" == "1" ]; then
+            upgrade_kubernetes_workers "$step"
+        fi
```

Comment:
I have to wonder - what happens if you resume an upgrade where the local primary had already been updated, but the remotes hadn't been?

(basically - are the `KUBERNETES_UPGRADE_LOCAL_PRIMARY` etc env vars updated each step? Otherwise, given a mix of 1.1 and 1.2 nodes upgrading to 1.4, would it end up only upgrading the 1.4 nodes?)

Reply:
> I have to wonder - what happens if you resume an upgrade where the local primary had already been updated, but the remotes hadn't been?

in that case KUBERNETES_UPGRADE_LOCAL_PRIMARY == 0 but KUBERNETES_UPGRADE_REMOTE_PRIMARIES and/or KUBERNETES_UPGRADE_SECONDARIES should equal 1. i borrowed the existing code.

these variables get set [here](https://github.com/replicatedhq/kURL/pull/4377/files#diff-3b622c4b740803afde003b07e5117e7f7d0836949348222439068ee6c9b98399R150)

[kubernetes_upgrade_discover_min_kubernetes_version](https://github.com/replicatedhq/kURL/pull/4377/files#diff-a1f3c001e0406a79e95cc96411af439600d0a27ce32abcdd6662013918ca4636R36) will find the lowest version of kubernetes on all nodes and start there.

tbh i havent yet tested this but i will

Reply:
you were correct. resume logic wasnt quite correct. i have fixed it